src/main/java/org/opensearch/neuralsearch/highlight/SemanticHighlightingResponseProcessor.java

```diff
@@ -0,0 +1,390 @@
+                    validHits.add(hit);
+                }
+            }
+
+            if (requests.isEmpty()) {
+                responseListener.onResponse(response);
+                return;
+            }
+
+            if (batchInference) {
+                mlClientAccessor.batchInferenceSentenceHighlighting(modelId, requests, ActionListener.wrap(batchResults -> {
+                    SearchResponse highlightedResponse = applyBatchHighlightResults(
+                        response,
+                        batchResults,
+                        validHits,
+                        semanticHighlightField
+                    );
+                    long totalTime = System.currentTimeMillis() - startTime;
+                    highlightedResponse = ProcessorUtils.updateResponseTookTime(highlightedResponse, totalTime);
+                    responseListener.onResponse(highlightedResponse);
+                }, error -> { handleProcessingError(error, response, responseListener); }));
+            } else {
+                processSequentialRequests(requests, validHits, 0, response, semanticHighlightField, responseListener, startTime);
+            }
+
+        } catch (Exception e) {
+            handleProcessingError(e, response, responseListener);
+        }
+    }
+
+    /**
+     * Handle errors with ignore failure logic
+     */
+    private void handleProcessingError(Exception error, SearchResponse response, ActionListener<SearchResponse> responseListener) {
+        if (ignoreFailure) {
+            log.warn(String.format(Locale.ROOT, "semantic highlighting failed but ignoring failure: %s", error.getMessage()));
+            responseListener.onResponse(response);
+        } else {
+            responseListener.onFailure(error);
+        }
+    }
+
+    /**
+     * Process requests sequentially - one at a time
+     */
+    private void processSequentialRequests(
+        List<SentenceHighlightingRequest> requests,
+        List<SearchHit> validHits,
+        int currentIndex,
+        SearchResponse response,
+        String semanticHighlightField,
+        ActionListener<SearchResponse> responseListener,
+        long startTime
+    ) {
+        // Base case: all requests processed
+        if (currentIndex >= requests.size()) {
+            long totalTime = System.currentTimeMillis() - startTime;
+            SearchResponse finalResponse = ProcessorUtils.updateResponseTookTime(response, totalTime);
+            responseListener.onResponse(finalResponse);
+            return;
+        }
+
+        // Process current request
+        mlClientAccessor.inferenceSentenceHighlighting(requests.get(currentIndex), ActionListener.wrap(highlightResults -> {
+            try {
+                applySingleHighlightResults(validHits.get(currentIndex), highlightResults, semanticHighlightField);
+                // Process next request recursively
```

Comment:
Can we avoid using a recursive approach here? For large results, it would consume stack memory unnecessarily.